rocAL/source/readers/webdataset_source_reader.cpp

```diff
@@ -0,0 +1,371 @@
+            if ((_sub_dir = opendir(_folder_path.c_str())) == nullptr)
+                THROW("WebDatasetSourceReader ShardID [" + TOSTR(_shard_id) + "] ERROR: Failed opening the directory at " + _folder_path);
+            _full_path = _folder_path;
+            while ((_entity = readdir(_sub_dir)) != nullptr) {
+                std::string entry_name(_entity->d_name);
+                if (strcmp(_entity->d_name, ".") == 0 || strcmp(_entity->d_name, "..") == 0)
+                    continue;
+                _index_name_list.push_back(entry_name);
+            }
+            std::sort(_index_name_list.begin(), _index_name_list.end());
+            // tar file path
+            if ((_sub_dir = opendir(_path.c_str())) == nullptr)
+                THROW("WebDatasetSourceReader ShardID [" + TOSTR(_shard_id) + "] ERROR: Failed opening the directory at " + _path);
+            std::string _full_path = _path;
+            while ((_entity = readdir(_sub_dir)) != nullptr) {
+                std::string entry_name(_entity->d_name);
+                if (strcmp(_entity->d_name, ".") == 0 || strcmp(_entity->d_name, "..") == 0)
+                    continue;
+                entry_name_list.push_back(entry_name);
+            }
+            std::sort(entry_name_list.begin(), entry_name_list.end());
+            _wds_shards.reserve(entry_name_list.size());
+            // Create n such std-streams for n paths
+            for (auto& path : entry_name_list)
+                _wds_shards.emplace_back(FileIOStream::open(_path + path));
+    }
+
+
+    // if (!_meta_data_reader) {
+        std::vector<SampleDescription> unfiltered_samples;
+        std::vector<ComponentDescription> unfiltered_components;
+
+        for (unsigned wds_shard_index = 0; wds_shard_index < entry_name_list.size(); ++wds_shard_index) {
+            unfiltered_samples.resize(0);
+            unfiltered_components.resize(0);
+            if (_index_paths.size() == 0)
+                parse_tar_files(unfiltered_samples, unfiltered_components, _wds_shards[wds_shard_index]);
+            else
+                parse_index_files(unfiltered_samples, unfiltered_components, _folder_path + _index_name_list[wds_shard_index]);
+
+            // After parsing add the contents to the map
+            for (auto& sample : unfiltered_samples) {
+                for (auto& component : sample.components) {
+                     if (webdataset_record_reader_from_components(component, wds_shard_index) != Reader::Status::OK)
+                                WRN("WebDatasetSourceReader ShardID [" + TOSTR(_shard_id) + "] WebDataset File reader cannot access the storage at " + _folder_path + component.filename);
+                }
+            }
+        }
+
+    if (!_file_names.empty())
+        LOG("WebDatasetSourceReader ShardID [" + TOSTR(_shard_id) + "] Total of " + TOSTR(_file_names.size()) + " images loaded from " + _full_path)
+    closedir(_sub_dir);
+    return ret;
+}
+
```

Comment:
Why is this redeclared?

Reply:
Not clear on the comment. Can you please elaborate this comment